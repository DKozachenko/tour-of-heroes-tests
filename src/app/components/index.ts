export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
