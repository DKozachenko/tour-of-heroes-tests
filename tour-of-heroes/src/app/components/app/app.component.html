<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
