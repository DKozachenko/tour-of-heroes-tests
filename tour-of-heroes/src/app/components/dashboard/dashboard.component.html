<h2 automation-id="heading" pw-automation-id="heading">Top Heroes</h2>
<div
  class="heroes-menu"
  automation-id="heroes-menu"
  pw-automation-id="heroes-menu"
>
  <a
    *ngFor="let hero of heroes"
    routerLink="/detail/{{ hero.id }}"
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
    {{ hero.name }}
  </a>
</div>

<app-hero-search
  automation-id="hero-search"
  pw-automation-id="hero-search"
></app-hero-search>
