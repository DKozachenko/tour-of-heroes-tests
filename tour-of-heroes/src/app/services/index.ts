export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
