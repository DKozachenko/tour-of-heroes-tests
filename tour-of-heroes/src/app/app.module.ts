import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
