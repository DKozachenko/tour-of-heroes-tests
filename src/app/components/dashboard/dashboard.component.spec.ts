import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { fakeAsync, tick } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { Location } from '@angular/common';
import { Router, RouterLink, RouterModule } from '@angular/router';
import {
  MockBuilder,
  MockRender,
  MockedComponentFixture,
  NG_MOCKS_ROOT_PROVIDERS,
} from 'ng-mocks';
import { of } from 'rxjs';
import { mock, instance, when, verify } from 'ts-mockito';
import { HeroService, HEROES } from '../../services';
import { DashboardComponent } from './dashboard.component';
import { AppModule } from '../../../app/app.module';

class PageObject {
  private fixtureDebugElement: DebugElement;

  private getElementByAutomationId(id: string): DebugElement {
    return this.fixtureDebugElement.query(By.css(`[automation-id=${id}]`));
  }

  private getElementsByAutomationId(id: string): DebugElement[] {
    return this.fixtureDebugElement.queryAll(By.css(`[automation-id=${id}]`));
  }

  constructor(fixture: MockedComponentFixture<DashboardComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get heading(): DebugElement {
    return this.getElementByAutomationId('heading');
  }

  get heroesMenu(): DebugElement {
    return this.getElementByAutomationId('heroes-menu');
  }

  get heroLinks(): DebugElement[] {
    return this.getElementsByAutomationId('hero-link');
  }

  get heroSearch(): DebugElement {
    return this.getElementByAutomationId('hero-search');
  }
}

describe('DashboardComponent', () => {
  let mockHeroService: HeroService;

  beforeEach(() => {
    mockHeroService = mock(HeroService);

    return MockBuilder(DashboardComponent, AppModule).mock(
      HeroService,
      instance(mockHeroService)
    );
  });

  function createFixture(): MockedComponentFixture<DashboardComponent> {
    return MockRender(DashboardComponent);
  }

  function createComponent(): DashboardComponent {
    return createFixture().point.componentInstance;
  }

  function mockCalls(): void {
    when(mockHeroService.getHeroes()).thenReturn(
      of(HEROES.slice(0, HEROES.length))
    );
  }

  describe('Functional', () => {
    describe('getHeroes', () => {
      it('should call "getHeroes" service method while creating', () => {
        mockCalls();
        createComponent();
        verify(mockHeroService.getHeroes()).once();
      });

      it('should call "getHeroes" service method via "getHeroes" component method', () => {
        mockCalls();
        const component = createComponent();
        component.getHeroes();
        verify(mockHeroService.getHeroes()).twice();
      });
    });

    it('should get 4 heroes (from 1 to 5 excluding) while creating', () => {
      mockCalls();
      const component = createComponent();
      const expectedHeroes = HEROES.slice(1, 5);

      expect(component.heroes).toHaveLength(4);
      expect(component.heroes).toStrictEqual(expectedHeroes);
    });
  });

  describe('Layout', () => {
    it('should contain heading with "Top Heroes" title', () => {
      mockCalls();
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.heading).not.toBeNull();
      expect(pageObject.heading.nativeElement.textContent).toBe('Top Heroes');
      expect(pageObject.heading.nativeElement).toMatchSnapshot();
    });

    it('should contain heroes list including 4 links with hero names and hero id as routerLink attribute', () => {
      mockCalls();
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.heroesMenu).not.toBeNull();
      expect(pageObject.heroesMenu.nativeElement).toMatchSnapshot();
      expect(pageObject.heroLinks).toHaveLength(4);

      const expectedHeroes = HEROES.slice(1, 5);
      for (let i = 0; i < pageObject.heroLinks.length; ++i) {
        const heroLink = pageObject.heroLinks[i];
        const hero = expectedHeroes[i];
        expect(heroLink.nativeElement.textContent).toContain(hero.name);
        expect(heroLink.injector.get(RouterLink).routerLink).toBe(
          `/detail/${hero.id}`
        );
        expect(heroLink.nativeElement).toMatchSnapshot();
      }
    });

    it('should contain "app-hero-search" component', () => {
      mockCalls();
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.heroSearch).not.toBeNull();
      expect(pageObject.heroSearch.nativeElement).toMatchSnapshot();
    });
  });
});

// https://ng-mocks.sudo.eu/guides/route/
describe('DashboardComponent:Routing', () => {
  let mockHeroService: HeroService;

  beforeEach(() => {
    mockHeroService = mock(HeroService);

    return MockBuilder(
      [
        DashboardComponent,
        RouterModule,
        RouterTestingModule.withRoutes([]),
        NG_MOCKS_ROOT_PROVIDERS,
      ],
      AppModule
    ).mock(HeroService, instance(mockHeroService));
  });

  function createFixture(): MockedComponentFixture<DashboardComponent> {
    return MockRender(DashboardComponent);
  }

  function mockCalls(): void {
    when(mockHeroService.getHeroes()).thenReturn(
      of(HEROES.slice(0, HEROES.length))
    );
  }

  it('should change location if appropriate hero link has clicked', fakeAsync(() => {
    mockCalls();
    const fixture = createFixture();
    const router = fixture.point.injector.get(Router);
    const location = fixture.point.injector.get(Location);

    if (fixture.ngZone) {
      fixture.ngZone.run(() => router.initialNavigation());
      tick();
    }

    const pageObject = new PageObject(fixture);
    for (let i = 0; i < pageObject.heroLinks.length; ++i) {
      const heroLink = pageObject.heroLinks[i];
      if (fixture.ngZone) {
        fixture.ngZone.run(() => {
          heroLink.triggerEventHandler('click', {
            button: 0,
          });
        });
        tick();
      }

      expect(location.path()).toBe(heroLink.injector.get(RouterLink).href);
    }
  }));
});
