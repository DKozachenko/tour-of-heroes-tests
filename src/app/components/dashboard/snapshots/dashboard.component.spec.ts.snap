// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardComponent Layout should contain "app-hero-search" component 1`] = `
<app-hero-search
  automation-id="hero-search"
  pw-automation-id="hero-search"
/>
`;

exports[`DashboardComponent Layout should contain heading with "Top Heroes" title 1`] = `
<h2
  automation-id="heading"
  pw-automation-id="heading"
>
  Top Heroes
</h2>
`;

exports[`DashboardComponent Layout should contain heroes list including 4 links with hero names and hero id as routerLink attribute 1`] = `
<div
  automation-id="heroes-menu"
  class="heroes-menu"
  pw-automation-id="heroes-menu"
>
  <a
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
     Bombasto 
  </a>
  
  <a
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
     Celeritas 
  </a>
  
  <a
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
     Magneta 
  </a>
  
  <a
    automation-id="hero-link"
    pw-automation-id="hero-link"
  >
     RubberMan 
  </a>
  
  
</div>
`;

exports[`DashboardComponent Layout should contain heroes list including 4 links with hero names and hero id as routerLink attribute 2`] = `
<a
  automation-id="hero-link"
  pw-automation-id="hero-link"
>
   Bombasto 
</a>
`;

exports[`DashboardComponent Layout should contain heroes list including 4 links with hero names and hero id as routerLink attribute 3`] = `
<a
  automation-id="hero-link"
  pw-automation-id="hero-link"
>
   Celeritas 
</a>
`;

exports[`DashboardComponent Layout should contain heroes list including 4 links with hero names and hero id as routerLink attribute 4`] = `
<a
  automation-id="hero-link"
  pw-automation-id="hero-link"
>
   Magneta 
</a>
`;

exports[`DashboardComponent Layout should contain heroes list including 4 links with hero names and hero id as routerLink attribute 5`] = `
<a
  automation-id="hero-link"
  pw-automation-id="hero-link"
>
   RubberMan 
</a>
`;
